hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/execution/OMGateway.java

```diff
@@ -0,0 +1,208 @@
+import org.apache.hadoop.ozone.om.OzoneManager;
+import org.apache.hadoop.ozone.om.OzoneManagerPrepareState;
+import org.apache.hadoop.ozone.om.exceptions.OMException;
+import org.apache.hadoop.ozone.om.lock.OmLockOpr;
+import org.apache.hadoop.ozone.om.ratis.execution.factory.OmRequestFactory;
+import org.apache.hadoop.ozone.om.ratis.execution.request.OMRequestBase;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMResponse;
+import org.apache.hadoop.ozone.protocolPB.OzoneManagerRequestHandler;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.hadoop.ipc.RpcConstants.DUMMY_CLIENT_ID;
+import static org.apache.hadoop.ipc.RpcConstants.INVALID_CALL_ID;
+
+/**
+ * entry for request execution.
+ */
+public class OMGateway {
+  private static final Logger LOG = LoggerFactory.getLogger(OMGateway.class);
+  private final LeaderRequestExecutor leaderExecutor;
+  private final OzoneManager om;
+  private final OmLockOpr omLockOpr;
+  private final AtomicLong requestInProgress = new AtomicLong(0);
+  private final AtomicLong uniqueIndex = new AtomicLong();
+
+  public OMGateway(OzoneManager om) throws IOException {
+    this.om = om;
+    this.omLockOpr = new OmLockOpr(om.getThreadNamePrefix());
+    this.leaderExecutor = new LeaderRequestExecutor(om, uniqueIndex);
+    if (om.isRatisEnabled()) {
+      om.getOmRatisServer().getOmStateMachine().getPostExecutor().registerIndexNotifier(this::indexNotifier);
+    } else {
+      uniqueIndex.set(PostExecutor.initLeaderIndex(om));
+    }
+  }
+
+  public void start() {
+    omLockOpr.start();
+    leaderExecutor.start();
+  }
+  public void stop() {
+    leaderExecutor.setEnabled(false);
+    leaderExecutor.stop();
+    omLockOpr.stop();
+  }
+
+  public OMResponse submit(OMRequest omRequest, byte[] clientId, long callId) throws ServiceException {
+    // TODO handle replay of request for external request
+    return submitInternal(omRequest, getClientId(clientId), getCallId(callId));
+  }
+
+  public OMResponse submit(OMRequest omRequest) throws ServiceException {
+    return submitInternal(omRequest, null, 0L);
+  }
+
+  public OMResponse submitInternal(OMRequest omRequest, String clientId, long callId) throws ServiceException {
+    requestInProgress.incrementAndGet();
+    RequestContext requestContext = new RequestContext();
+    requestContext.setRequest(omRequest);
+    requestContext.setFuture(new CompletableFuture<>());
+    CompletableFuture<OMResponse> f = requestContext.getFuture().whenComplete(
+        (r, th) -> handleAfterExecution(requestContext, th));
+    OmLockOpr.OmLockInfo lockInfo = null;
+    try {
+      OMRequestBase requestBase = OmRequestFactory.createClientRequest(omRequest, om);
```

Comment:
Reusing Request name seemed wrong in the original code itself. Request == a bunch of attributes (nouns) and a verb for what needs to be done. These `ClientRequest`s embed the entire logic for executing the code. They in essense capture the lamba to be performed for each request.
```suggestion
      OMRequestBase requestBase = OmRequestFactory.createRequestExecutor(omRequest, om);
```

Reply:
or `createRequestProcessors`